src/graph/query_graph.c

```diff
@@ -182,27 +186,56 @@ QueryGraph* QueryGraph_New(size_t node_cap, size_t edge_cap) {
     return g;
 }
 
-void BuildQueryGraph(const GraphContext *gc, QueryGraph *qg, Vector *entities) {    
+void BuildQueryGraph(const GraphContext *gc, QueryGraph *qg, Vector *entities) {
+    /* Store node pointers to reduce the number of lookups in edge construction.
+     * Only indices corresponding to real nodes will be valid references, and pointers
+     * may appear repeatedly. */
+    Node *nodes[Vector_Size(entities)];
+
     /* Introduce nodes first. */
     for(int i = 0; i < Vector_Size(entities); i++) {
         AST_GraphEntity *entity;
         Vector_Get(entities, i, &entity);
         if(entity->t != N_ENTITY) continue;
-        _BuildQueryGraphAddNode(gc, entity, qg);
+        nodes[i] = _BuildQueryGraphAddNode(gc, entity, qg);
     }
 
     /* Introduce edges. */
     for(int i = 0; i < Vector_Size(entities); i++) {
         AST_GraphEntity *entity;
         Vector_Get(entities, i, &entity);
         if(entity->t != N_LINK) continue;
-        AST_GraphEntity *l_entity;
-        AST_GraphEntity *r_entity;
-        Vector_Get(entities, i-1, &l_entity);
-        Vector_Get(entities, i+1, &r_entity);
+        _BuildQueryGraphAddEdge(gc, (AST_LinkEntity*)entity, nodes[i-1], nodes[i+1], qg);
+    }
+}
+
+void QueryGraph_AddNode(QueryGraph *g, Node *n, char *alias) {
+    _QueryGraph_AddEntity((GraphEntity*)n,
+    alias,
+    (GraphEntity ***)&g->nodes,
+    &g->node_aliases,
+    &g->node_count,
+    &g->node_cap);
+}
 
-        _BuildQueryGraphAddEdge(gc, entity, l_entity, r_entity, qg);
+Node** QueryGraph_ConnectedComponents(const QueryGraph *qg, int *component_count) {
```

Comment:
I think it would be better if QueryGraph_ConnectedComponents were to return QueryGraphs
each one containing just a single connected component.

Reply:
Huh, that would save a lot of effort in building the expression trees. I like that idea a lot!